### DISCUSSION TITLE
웹서버 vs WAS

### DISCUSSION THREAD

**Discussion by byhhh2** [woowacourse-study/2022-woowahan-bansanghwe#99] [Network]
![image](https://user-images.githubusercontent.com/52737532/192161421-2f896606-9310-4b80-8d1b-1ff2d2b375e6.png)

- Web Server
- WAS : Web Application Server

## 🍪 Web Server란?

- **하드웨어로써** : web server가 설치되어 있는 컴퓨터
- **소프트웨어로써** : 웹 브라우저(클라이언트)로부터 HTTP 요청을 받아 정적인 컨텐츠를 (.html, .jpg, .css 등) 제공하는 프로그램 (즉, 클라이언트의 요청을 서비스한다.)
- HTTP Server라고도 불린다. 

### Web Server의 기능

> 요청에 따라 기능이 달라진다.
> 
- 정적인 컨텐츠 요청 : WAS를 거치지 않고 자원 제공
- 동적인 컨텐츠 요청 : 클라이언트의 요청을 WAS에 보내고, WAS가 처리한 결과물을 클라이언트에게 응답으로 준다.
- Web Container 혹은 Servlet Container 라고도 불린다.
    - Container : JSP, Servlet을 실행시킬 수 있는 소프트웨어 (즉, 구동환경)
- SSL에 대한 암복호화 처리

> **정적인 컨텐츠** 
>
> - Web Server가 경로 이름을 받아와 경로와 일치하는 컨텐츠를 반환
> - 항상 동일한 페이지를 반환
> - ex) html, image, css, js
> 
> **동적인 컨텐츠** 
> 
> - 인자에 맞게 동적인 컨텐츠를 반환 
> - WAS에 의해 실행되는 프로그램을 통해 만들어진 결과물 
> - ex) Java program (sevlet(WAS) 위에서 실행), asp, php, jsp
> 

### Web Server의 예시

- Apache, Nginx 등

## 🍪 WAS(Web Application Server)란?

- 동적인 컨텐츠를 제공하기 위해 만들어진 server
    - 동적인 컨텐츠 : DB 조회나 다양한 로직 처리를 요구한다.
- HTTP를 통해 컴퓨터 어플리케이션을 수행해주는 미들웨어이다.

### WAS의 기능

- WAS = Web Server + Web Container
- Web Server의 기능을 분리하려 처리하고자 하는 목적으로 제시되었다.
- DB 조회, 비즈니스 로직 수행 등
- 파일을 수행한 결과를 다시 Web Server로 보내주는 역할을 한다.

### WAS의 예시

- Tomcat, JBoss, Jeus, Web Sphere

## 🍪 Web Server와 WAS는 왜 나누어져 있는가?

클라어언트의 요청으로 HTML문서와, 이미지들을 응답으로 보내주어야 할 때 HTML문서와 이미지는 함께 보내지는 것이 아니며, HTML 문서를 먼저 받고 그에 맞게 필요한 이미지들을 서버에 다시 요청한다. 이때 Web Server가 존재하기 때문에 WAS까지 가지 않고 빠르게 이미지들을 응답할 수 있다. 즉, Web Server와 WAS가 나누어져 있기 때문에 Web Server에서는 정적 컨텐츠만 처리하도록, WAS는 동적 컨텐츠를 처리하도록 하여 서버의 부담을 감소시킨다. 

만약 동적인 컨텐츠가 요청되었을 때 Web Server만이 존재한다면 사용자가 원하는 요청에 대한 결과값을 모두 미리 만들어놓고 서비스를 해야한다. 이는 자원의 한계때문에 불가능하다고 볼 수 있다. 이때 WAS를 두어 요청에 맞는 데이터를 직접 가져와 로직에 따라 처리하기 때문에 자원을 효율적으로 사용할 수 있다. 

또, 서버가 여러개 사용되므로써 무중단 운영을 위한 장애 극복도 쉽게 대응할 수 있는데, 만약 하나의 서버에서 오류가 발생했을 지라도 나머지 서버가 대응하면 되므로 사용자는 오류를 느끼지 못한채 서비스를 이용할 수 있다. 

> 결론?

- 자원 이용의 효율성
- 장애 극복 
- 유지보수의 편의성
> 

## 참고

- [https://gmlwjd9405.github.io/2018/10/27/webserver-vs-was.html](https://gmlwjd9405.github.io/2018/10/27/webserver-vs-was.html)
- [https://codechasseur.tistory.com/25](https://codechasseur.tistory.com/25)

**Comment by al-bur** [comment]
웹 서버란 무엇인가요?

**Comment by al-bur** [comment]
WAS는 무엇인가요?

**Comment by al-bur** [comment]
사용해보신 WAS?

**Comment by al-bur** [comment]
사용해보신 WS?

**Comment by al-bur** [comment]
nginx에 어떤 설정을 해주셨나요?

**Comment by al-bur** [comment]
WAS를 통해서는 정적인 컨텐츠를 제공해줄 수 없나요?

**Reply by al-bur** [reply]
왜 WS를 사용하는 것이 성능적으로, 메모리 절약적으로 좋나요?

**Comment by al-bur** [comment]
프로젝트에서 WS/WAS가 어떻게 구성되어있는지?

**Comment by al-bur** [comment]
프로젝트에서 CI/CD를 구축하신건가요?

**Comment by al-bur** [comment]
CI/CD를 왜 구축하시게 된거죠?

**Comment by al-bur** [comment]
어떤 테스트 작성하셨나요?

**Comment by al-bur** [comment]
스타일을 상속받아서 해준 장단점은?

**Comment by airman5573** [comment]
web server 와 was는 어떻게 통신을 하나요?

**Comment by airman5573** [comment]
써보신 ws는?

**Comment by airman5573** [comment]
nginx를 사용할때 어려웠던점은?

**Comment by airman5573** [comment]
ec2에 어떻게 접근하셨나요?

**Comment by airman5573** [comment]
어떤 설정을 해주셨나요?

**Comment by airman5573** [comment]
ws가 꼭 필요한가요?

**Comment by airman5573** [comment]
어떤 측면에서 ws를 쓰는게 메모리적으로 좋은지?

**Comment by airman5573** [comment]
was를 더 앞에 두면 되는거 아닌가요?

**Comment by airman5573** [comment]
보안적으로 어떻게 유리한거죠?

**Comment by airman5573** [comment]
ws가 없을때 서버가 어떤 보안 위협을 받을 수 있을까요?

**Comment by airman5573** [comment]
ec2를 사용하시면서 어려웠던 부분은?

**Comment by airman5573** [comment]
react-testing-library를 쓰면서 어려웠던 점은?

**Comment by airman5573** [comment]
storybook/testing-library를 왜 쓰셨나요?

**Comment by airman5573** [comment]
react-test-renderer는 어떤 패키지인가요?

**Comment by airman5573** [comment]
type-fest 패키지는 어떤 용도로 설치하신건가요?

**Comment by airman5573** [comment]
컴포넌트 작성시의 컨벤션이 있나요?

**Comment by airman5573** [comment]
props를 받는것은 어떤가요?

**Comment by airman5573** [comment]
스타일링을 할때 상속받는것의 단점은 (자유로움의 단점)
